src/core/DY.Core.Trace.SplitUtils.fst

```diff
@@ -0,0 +1,285 @@
+      fmap_trace_entry f e == e' /\
+      fmap_trace f tr2 == tr2'
+    )
+  = norm_spec [zeta; delta_only [`%trace_split]] (trace_split #a);
+    norm_spec [zeta; delta_only [`%trace_split]] (trace_split #b);
+    let Snoc hd entry = tr in
+    if ts = last_ts tr then ()
+    else trace_split_fmap hd ts f
+
+(*** Trace arithmetic ***)
+
+/// We define trace concatenation (the monoid operation), as well as
+/// trace subtraction, which is defined for traces tr2 <$ tr1, and yields the suffix of
+/// tr1 beginning after tr2.
+
+[@@"opaque_to_smt"]
+let rec trace_concat (#label_t:Type) (tr1 tr2:trace_ label_t)
+  : trace_ label_t
+  = match tr2 with
+    | Nil -> tr1
+    | Snoc hd e -> Snoc (trace_concat tr1 hd) e
+
+[@@"opaque_to_smt"]
+let rec trace_subtract (#label_t:Type) (tr1:trace_ label_t) (tr2:trace_ label_t{tr2 <$ tr1})
+  : trace_ label_t
+  = if length tr1 = length tr2 then Nil
+    else begin
+      let Snoc hd e = tr1 in
+      grows_cases tr2 tr1;
+      Snoc (trace_subtract hd tr2) e
+    end
+
+let (<++>) = trace_concat
+let (<-->) = trace_subtract
+
+/// Traces form a monoid with trace concatenation as operation, and Nil as unit.
+
+let rec trace_concat_nil (#label_t:Type) (tr:trace_ label_t)
+  : Lemma
+    (
+      tr <++> Nil == tr /\
+      Nil <++> tr == tr
+    )
+    [SMTPatOr [[SMTPat (tr <++> Nil)]; [SMTPat (Nil <++> tr)]]]
+  = norm_spec [zeta; delta_only [`%trace_concat]] (trace_concat #label_t);
+    match tr with
+    | Nil -> ()
+    | Snoc hd _ -> trace_concat_nil hd
+
+let rec trace_concat_assoc (#label_t:Type) (tr1 tr2 tr3:trace_ label_t)
+  : Lemma
+    (((tr1 <++> tr2) <++> tr3) == (tr1 <++> (tr2 <++> tr3)))
+  = norm_spec [zeta; delta_only [`%trace_concat]] (trace_concat #label_t);
+    match tr3 with
+    | Nil -> ()
+    | Snoc hd _ -> trace_concat_assoc tr1 tr2 hd
+
+let rec trace_subtract_nil (#label_t:Type) (tr:trace_ label_t)
+  : Lemma
+    ((tr <--> Nil) == tr)
+    [SMTPat (tr <--> Nil)]
+  = norm_spec [zeta; delta_only [`%trace_subtract]] (trace_subtract #label_t);
+    match tr with
+    | Nil -> ()
+    | Snoc hd _ -> trace_subtract_nil hd
+
+let trace_subtract_snoc_left (#label_t:Type) (tr1 tr2:trace_ label_t) (e:trace_entry_ label_t)
+  : Lemma
+    (requires tr1 <$ tr2)
+    (ensures ((Snoc tr2 e) <--> tr1) == Snoc (tr2 <--> tr1) e)
+  = norm_spec [zeta; delta_only [`%trace_subtract]] (trace_subtract #label_t)
+
+let trace_subtract_snoc_left' (#label_t:Type) (tr1 tr2:trace_ label_t)
+  : Lemma
+    (requires Snoc? tr2 /\ (let Snoc hd _ = tr2 in tr1 <$ hd))
+    (ensures (
+      let Snoc hd e = tr2 in
+      (tr2 <--> tr1) == Snoc (hd <--> tr1) e
+    ))
+  = norm_spec [zeta; delta_only [`%trace_subtract]] (trace_subtract #label_t)
+
+let trace_concat_snoc_right (#label_t:Type) (tr1 tr2:trace_ label_t) (e:trace_entry_ label_t)
+  : Lemma
+    (tr1 <++> (Snoc tr2 e) == Snoc (tr1 <++> tr2) e)
+  = norm_spec [zeta; delta_only [`%trace_concat]] (trace_concat #label_t)
```

Comment:
This can be derived from the lemma below
```fstar
let trace_concat_snoc_right (#label_t:Type) (tr1 tr2:trace_ label_t) (e:trace_entry_ label_t)
  : Lemma
    (tr1 <++> (Snoc tr2 e) == Snoc (tr1 <++> tr2) e)
  = snoc_is_concat_singleton tr2 e;
    snoc_is_concat_singleton (tr1 <++> tr2) e;
    trace_concat_assoc tr1 tr2 (Snoc Nil e)
```

Reply:
With the appropriate SMTPat on `snoc_is_concat_singleton`, it seems possible to remove this lemma altogether:
`[SMTPat (tr <++> (Snoc Nil e))]` is sufficient to trigger `snoc_is_concat_singleton` where `trace_concat_snoc_right` is currently used, though it does feel a bit odd to rely on. I guess I will move `trace_concat_snoc_right` into the derived section for now, since we can get by without it in the core section.